testing/matrix-sdk-integration-testing/src/tests/e2ee.rs

```diff
@@ -4,20 +4,241 @@ use std::{
+
+    let alice_sas =
+        alice_verification_request.start_sas().await?.expect("must have a sas verification");
+
+    bob.sync_once().await?;
+    let bob_sas = assert_matches!(bob_verification_request.state(), VerificationRequestState::Transitioned { verification: Verification::SasV1(bob_sas) } => bob_sas);
+
+    assert_matches!(alice_sas.state(), SasState::Started { .. });
+    assert_matches!(bob_sas.state(), SasState::Started { .. });
+
+    bob_sas.accept().await?;
+    assert_matches!(bob_sas.state(), SasState::Accepted { .. });
+    alice.sync_once().await?;
+    assert_matches!(alice_sas.state(), SasState::Accepted { .. });
+
+    assert!(!alice_sas.can_be_presented());
+    assert!(!bob_sas.can_be_presented());
+
+    // Let a little crypto messages dance happen.
+    alice.sync_once().await?;
+    bob.sync_once().await?;
+    assert_matches!(alice_sas.state(), SasState::Accepted { .. });
+    assert_matches!(bob_sas.state(), SasState::KeysExchanged { .. });
+
+    alice.sync_once().await?;
+    let alice_emojis =
+        assert_matches!(alice_sas.state(), SasState::KeysExchanged { emojis, .. } => emojis)
+            .expect("alice received emojis");
+    let bob_emojis =
+        assert_matches!(bob_sas.state(), SasState::KeysExchanged { emojis, .. } => emojis)
+            .expect("bob received emojis");
+
+    assert!(alice_sas.can_be_presented());
+    assert!(bob_sas.can_be_presented());
+
+    assert_eq!(format_emojis(alice_emojis.emojis), format_emojis(bob_emojis.emojis));
+
+    alice_sas.confirm().await?;
+    bob_sas.confirm().await?;
+    assert_matches!(alice_sas.state(), SasState::Confirmed);
+    assert_matches!(bob_sas.state(), SasState::Confirmed);
+
+    // Moar crypto dancing.
+    alice.sync_once().await?;
+    bob.sync_once().await?;
+    assert_matches!(alice_sas.state(), SasState::Confirmed);
+    assert_matches!(bob_sas.state(), SasState::Done { .. });
+
+    alice.sync_once().await?;
+    assert_matches!(alice_sas.state(), SasState::Done { .. });
+    assert_matches!(bob_sas.state(), SasState::Done { .. });
+
+    // Wait for remote echos for verification status requests.
+    alice.sync_once().await?;
+    bob.sync_once().await?;
+
+    // Both users appear as verified to each other.
+    let alice_bob_ident =
+        alice.encryption().get_user_identity(bob.user_id().unwrap()).await?.unwrap();
+    assert!(alice_bob_ident.is_verified());
+
+    let bob_alice_ident =
+        bob.encryption().get_user_identity(alice.user_id().unwrap()).await?.unwrap();
+    assert!(bob_alice_ident.is_verified());
+
+    // Both user devices appear as verified to the other user.
+    let alice_bob_device = alice_sas.other_device();
+    assert_eq!(alice_bob_device.user_id(), bob.user_id().unwrap());
+    assert_eq!(alice_bob_device.device_id(), bob.device_id().unwrap());
+    assert_eq!(alice_bob_device.local_trust_state(), LocalTrust::Unset);
```

Comment:
The local trust check probably belong into a separate test, but I guess the setup is too annoying.